codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -35,133 +32,79 @@
 import java.util.NoSuchElementException;
 import java.util.Set;
 import java.util.TimeZone;
-import java.util.TreeSet;
 
-import static io.netty.util.internal.ObjectUtil.*;
+import static io.netty.util.HashingStrategy.JAVA_HASHER;
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
 
+/**
+ * Default implementation of {@link Headers};
+ *
+ * @param <T> the type of the header name and value.
+ */
 public class DefaultHeaders<T> implements Headers<T> {
     /**
-     * Allows users of this interface to specify a hash code other than the default {@link Object#hashCode()}
-     */
-    public interface HashCodeGenerator<T> {
-        /**
-         * Obtain the hash code for the given {@code name}
-         *
-         * @param name The name to generate the hash code for
-         * @return The hash code for {@code name}
-         */
-        int generateHashCode(T name);
-    }
-
-    /**
-     * Allows users to convert the {@code name} elements before being processed by this map
-     */
-    public interface NameConverter<T> {
-        /**
-         * Convert the {@code name} to some other form of the same object type
-         *
-         * @param name The object to convert
-         * @return The results of the conversion
-         */
-        T convertName(T name);
-    }
-
-    /**
-     * Uses the {@link #hashCode()} method to generate the hash code.
+     * How big the underlying array is for the hash data structure.
+     * <p>
+     * This should be a power of 2 so the {@link #index(int)} method can full address the memory.
      */
-    public static final class JavaHashCodeGenerator<T> implements HashCodeGenerator<T> {
-        @Override
-        public int generateHashCode(T name) {
-            return name.hashCode();
-        }
-    }
+    private static final int ARRAY_SIZE = 1 << 4;
```

Comment:
@Scottmitch @normanmaurer can we play around with the size of this a bit? I mean in those benchmarks with wrk the header size = 3 so 16 is plenty, but what if we have bigger headers? Also this used to be a prime. I think we should gather some empiric evidence whether prime vs. power of two does anything. Further, we should not be afraid to size this array a little too big, say 32 if power of two or even bigger if we choose a prime. The "wasted" memory for small headers is minimal, but for bigger headers the perf boost might well be worth it.